File: Networks/listen_and_accept.py
import socket
import threading

def handle_connection(conn):
    while True:
        # data = conn.recv(1024): This receives data from the client connection conn in chunks of up to 1024
        # bytes and assigns it to the variable data.
        data = conn.recv(1024)
        if not data:
            break
        print(f"Received: {data.decode('utf-8')}")
    conn.close()

def server():
    # Create a TCP socket
    # with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as server_socket::
    # This creates a TCP socket server_socket using IPv4 (AF_INET) and the TCP protocol (SOCK_STREAM).
    # The with statement ensures the socket is closed properly after use.
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as server_socket:
        # Bind the socket to localhost and a random port
        # server_socket.bind(('127.0.0.1', 0)): This binds the server socket to the localhost (127.0.0.1)
        # and a random port. The 0 indicates that the OS should choose an available port.
        server_socket.bind(('127.0.0.1', 3000))
        # server_socket.listen(5): This sets the server socket to listen for incoming connections with a
        # maximum queue size of 5 pending connections.
        # Listen for incoming connections
        server_socket.listen(5)
        print(f"Server listening on {server_socket.getsockname()}")

        # Accept incoming connections in a loop
        while True:
            conn, addr = server_socket.accept()
            print(f"Accepted connection from {addr}")
            # Start a new thread to handle the connection
            threading.Thread(target=handle_connection, args=(conn,), daemon=True).start()

def client(server_address):
    # Create a TCP socket
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as client_socket:
        # Connect to the server
        client_socket.connect(server_address)
        print(f"Connected to server at {server_address}")
        # data = "Hello, server!"
        # client_socket.sendall(data.encode('utf-8'))
        # Close the connection immediately to initiate graceful termination
        client_socket.close()

# Start the server in a separate thread
server_thread = threading.Thread(target=server, daemon=True)
server_thread.start()

# Wait for a short time to ensure the server is up and running
import time
time.sleep(1)

# Get the server address from the server thread
server_address = server()

# Start the client
client(server_address)

File: Networks/listen_and_accept.go
package ch03

import (
    "io"
    "net"
    "testing"
)

func TestDial(t *testing.T) {
    // Create a listener on the IP address 127.0.0.1 with a random port.
    listener, err := net.Listen("tcp", "127.0.0.1:")
    if err != nil {
        t.Fatal(err)
    }
    done := make(chan struct{})

    // Goroutine to accept incoming connections
		// In the provided code, this goroutine is responsible for continuously accepting incoming TCP connections
		// from clients. It's created using the go keyword followed by an anonymous function func() { ... }.
		// Inside this anonymous function, the logic for accepting incoming connections is implemented.
    go func() {
        // Defer sending a signal to indicate goroutine completion
        defer func() { done <- struct{}{} }()

        // Loop to accept incoming TCP connections
        for {
            // Accept incoming connections
            conn, err := listener.Accept()
            if err != nil {
                // Log any errors but continue listening for connections
                t.Log(err)
                return
            }

            // Goroutine to handle each connection
            go func(c net.Conn) {
                // Defer closing the connection and sending a signal to indicate goroutine completion
                defer func() {
                    c.Close()
                    done <- struct{}{}
                }()

                // Buffer to read data from the connection
                buf := make([]byte, 1024)

                // Loop to read data from the connection
                for {
                    // Read data from the connection into the buffer
                    n, err := c.Read(buf)
                    if err != nil {
                        // If an error occurred while reading, log it
                        if err != io.EOF {
                            t.Error(err)
                        }
                        return
                    }
                    // Log the received data
                    t.Logf("received: %q", buf[:n])
                }
            }(conn)
        }
    }()

    // Dial the listener to establish a connection
    conn, err := net.Dial("tcp", listener.Addr().String())
    if err != nil {
        // If an error occurred while dialing, fail the test
        t.Fatal(err)
    }
    // Close the connection to initiate graceful termination
    conn.Close()

    // Wait for the goroutines to finish
    <-done
    // Close the listener
    listener.Close()
    // Wait for the listener goroutine to finish
    <-done
}
